tcms/rpc/tests/test_testcase.py

```diff
@@ -603,3 +736,59 @@ def verify_api_with_permission(self):
     def verify_api_without_permission(self):
         with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
             self.rpc_client.TestCase.comments(self.case.pk)
+
+
+class TestAddAttachmentPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.add_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.case = TestCaseFactory()
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.add_attachment(self.case.pk,
+                                                'test.txt',
+                                                'test')
+        attachments = Attachment.objects.attachments_for_object(self.case)
+        self.assertEqual(len(attachments), 1)
+        self.assertEqual(attachments[0].object_id, str(self.case.pk))
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.add_attachment(self.case.pk,
+                                                    'test.txt',
+                                                    'test')
+
+
+class TestListAttachments(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.case = TestCaseFactory()
+
+    def test_list_attachments(self):
+        self.rpc_client.TestCase.add_attachment(self.case.pk, 'attachment.txt', 'a2l3aXRjbXM=')
+        attachments = self.rpc_client.TestCase.list_attachments(self.case.pk)
+        self.assertEqual(len(attachments), 1)
+
+    def test_list_attachments_with_wrong_case_id(self):
+        with self.assertRaisesRegex(Fault, 'TestCase matching query does not exist'):
+            self.rpc_client.TestCase.list_attachments(-1)
+
+
+class TestListAttachmentPermissions(APIPermissionsTestCase):
```

Comment:
You can merge TestListAttachments class with this one and validate both the permissions and the functionality in one place. For simpler methods that works nicely. 

If you don't want to change it it can stay like that as well.

Note the missing scenario for non existing TC id! In case you aren't changing this class then add it to TestListAttachments.

Reply:
I took the reference from `tcms/rpc/tests/test_testplan.py` to implement tests for `list_attachments`. It seems like there is no easy way to add an attachment to a `TestCase` or a `TestPlan`. 

The sole reason for having two separate tests one for permissions and other for testing the functionality is the requirement of `attachments.add_attachment` permissions for adding an attachment which is not feasible in an instance of `APIPermissionsTestCase` class.